### ISSUE TITLE
Range Picker - allow presets to use current time as of user clicking the option

### ISSUE THREAD

**Issue by revoyrm** [ant-design/ant-design#43203]
### What problem does this feature solve?

We use presets for things like "Last Hour" but with the current implementation of presets, they are only calculated on page load/component mount. So, if the user waits on a page for a couple of minutes before pressing "Last Hour", the times chosen are based on the last hour from when the user entered the page and not based on when the user actually selects the option. I can reproduce using your preset example https://ant.design/components/date-picker#components-date-picker-demo-presetted-ranges by 
1. using the preset range picker
2.  waiting while  on the page for a minute or 2
3. choosing the last 7 days in the date and Time box. 
4. Result -> the end time will be when I loaded the page, not when I chose the preset.

### What does the proposed API look like?

Possibility 1: the presets prop could take a callback that takes the preset label and returns the start and end date array. This allows us to calculate the value on the preset being chosen rather than the component mounting.

Possibility 2: the onChange handler could pass the preset label as an optional parameter that would allow us to recalculate the date and time array at the point of it being selected.

![image](https://github.com/ant-design/ant-design/assets/11396904/7fd4ff87-a72d-49d7-802c-0d11c553eee1)

One possible work around for anyone looking to do this: Recalculate the  presets when the user click's the input. (onClick prop).  Would still be nice to have the ability to do this built in but this does work for our needs at the moment.

**Comment by Wxh16144** [comment]
That's a great suggestion, I had a similar requirement in my team a long time ago, but I forgot how I solved it then, I'll have to review the code 😅

**Comment by MadCcc** [comment]
`value` support function would be okay.

**Comment by MadCcc** [comment]
Released in `5.8.0`

**Comment by motaylormo** [comment]
If it's released now, what's the syntax to use it?
